cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -197,7 +199,7 @@ class RowGroupSkipper {
       const auto row_group = metadata_->RowGroup(row_group_idx);
 
       const auto num_rows = row_group->num_rows();
-      if (CanSkip(*row_group)) {
+      if (IsExplicitlySkipped(row_group_idx) || CanSkip(*row_group)) {
```

Comment:
Can we use a std::vector<int> instead? Otherwise we loop N times to skip N-1, thus a O(n^2) worse case if we have 1 row group per fragment. We can preserve this loop by breaking the assumption that `row_group_idx`  is a row group id, e.g. if the vector is empty, use the value, otherwise use `row_group_set_[row_group_idx_]`, `IsExplicitlySkipped` would disappear.